src/ray/raylet/local_object_manager.h

```diff
@@ -171,6 +204,29 @@ class LocalObjectManager {
   /// This class is accessed by both the raylet and plasma store threads. The
   /// mutex protects private members that relate to object spilling.
   mutable absl::Mutex mutex_;
+
+  ///
+  /// Fields below are used to delete spilled objects.
+  ///
+
+  /// A list of object id and url pairs that need to be deleted.
+  /// We don't instantly delete objects when it goes out of scope from external storages
+  /// because those objects could be still in progress of spilling.
+  std::queue<ObjectID> spilled_object_pending_delete_ GUARDED_BY(mutex_);
+
+  /// Mapping from object id to url_with_offsets. We cannot reuse pinned_objects_ because
+  /// pinned_objects_ entries are deleted when spilling happens.
+  absl::flat_hash_map<ObjectID, std::string> spilled_objects_url_ GUARDED_BY(mutex_);
```

Comment:
Does this work in the multi node case?

Reply:
Let me add a multi-node examples. But the current strategy is that each raylet only takes care of deletion of primary copy.

Reply:
I see, so all the objects in the file will be primary copies? Maybe add a comment on these data structures as to why they work and don't need to be synced globally.

Reply:
Yeah we only spill primary copies (spilling only happens from objects inside the `pinned_objects_` field) from the node that has the primary copies, and restored objects were never a primary copy again. cc @stephanie-wang just in case I am missing something here.